pkg/webhook/v1alpha2/applicationconfiguration/mutating_handler.go

```diff
@@ -0,0 +1,174 @@
+package applicationconfiguration
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	"github.com/davecgh/go-spew/spew"
+	crdv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
+	"github.com/crossplane/oam-kubernetes-runtime/pkg/oam"
+	"github.com/crossplane/oam-kubernetes-runtime/pkg/oam/util"
+)
+
+const (
+	// TraitTypeField is the special field indicate the type of the traitDefinition
+	TraitTypeField = "name"
+	// TraitSpecField indicate the spec of the trait in ApplicationConfiguration
+	TraitSpecField = "properties"
+)
+
+// MutatingHandler handles Component
+type MutatingHandler struct {
+	Client client.Client
+
+	// Decoder decodes objects
+	Decoder *admission.Decoder
+}
+
+// log is for logging in this package.
+var mutatelog = logf.Log.WithName("applicationconfiguration mutate webhook")
+
+var _ admission.Handler = &MutatingHandler{}
+
+// Handle handles admission requests.
+func (h *MutatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &v1alpha2.ApplicationConfiguration{}
+
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	// mutate the object
+	if err := h.Mutate(obj); err != nil {
+		mutatelog.Error(err, "failed to mutate the applicationConfiguration", "name", obj.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	mutatelog.Info("Print the mutated obj", "obj name", obj.Name, "mutated obj", spew.Sdump(obj.Spec))
+
+	marshalled, err := json.Marshal(obj)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	resp := admission.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, marshalled)
+	if len(resp.Patches) > 0 {
+		mutatelog.Info("admit ApplicationConfiguration",
+			"namespace", obj.Namespace, "name", obj.Name, "patches", util.JSONMarshal(resp.Patches))
+	}
+	return resp
+}
+
+// Mutate sets all the default value for the Component
+func (h *MutatingHandler) Mutate(obj *v1alpha2.ApplicationConfiguration) error {
+	mutatelog.Info("mutate", "name", obj.Name)
+
+	for compIdx, comp := range obj.Spec.Components {
+		var updated bool
```

Comment:
do we really need this?  Can't we just blindly copy back no matter what? The logic will be much simplified

Reply:
I don't think add this flag makes any complex here. I don't feel bindly copy back is good.